### PR TITLE
Add an ssh tunnel option to the /proxy endpoint

### PR DESCRIPTION
**Pull request by cjcullen**
Key generation and propagation is handled on gce.
Tunnels are recycled after 5 minutes.
The ssh-proxy functionality is not enabled by default for the gce provider.

I can squash commits when this gets ready to merge.

@brendandburns

### CODE REVIEW

**Comment by quinton-hoole** [comment]
This is a lot of new code.  What issue does it address?  Is it v1.0?

**Comment by alex-mohr** [comment]
I think part of #3168 ?

**Comment by cjcullen** [comment]
This addresses #3168 for the /proxy case (we won't need kubelet listening on the public internet for proxy).

File: pkg/util/ssh.go [hunk 1]
```diff
@@ -172,3 +153,95 @@ func MakePrivateKeySigner(key string) (ssh.Signer, error) {
...
+func (l SSHTunnelList) Open() error {
+	for ix := range l {
+		if err := l[ix].Tunnel.Open(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (l SSHTunnelList) Close() {
```
Comment (brendandburns) [inline pkg/util/ssh.go]: Document that this is async, and tunnels won't be closed when the method returns.

File: pkg/util/util.go [hunk 2]
```diff
@@ -508,3 +508,16 @@ func GetClient(req *http.Request) string {
 	}
 	return "unknown"
 }
+
+func FileExists(filename string) (bool, error) {
+	file, err := os.Open(filename)
```
Comment (brendandburns) [inline pkg/util/util.go]: in a different context, it was pointed out to me that you could use ```os.Stat``` here and save yourself having to close the file.

**Comment by brendandburns** [comment]
Generally LGTM, needs a rebase (and tests, if you felt like writing them...)

**Comment by kprobst** [comment]
Mostly LGTM, a couple of nits.

**Comment by cjcullen** [comment]
Note: this is not enabled by default. On GCE, it will only be enabled if grains.proxy_ssh_user is set.

**Comment by quinton-hoole** [comment]
@cjcullen and @brendanburns are working on the missing tests, which can come in a separate PR. This PR LGTM.

**Pull request merged**